### PR TITLE
Split raw locations

### PR DESCRIPTION
**Pull request by psibi**
Supersedes https://github.com/commercialhaskell/stack/pull/4422

### CODE REVIEW

File: subs/pantry/src/Pantry/Types.hs [hunk 1]
```diff
@@ -179,7 +194,7 @@ data PantryConfig = PantryConfig
   -- ^ Want to try updating the index once during a single run for missing
   -- package identifiers. We also want to ensure we only update once at a
   -- time. Start at @True@.
-  , pcParsedCabalFilesImmutable :: !(IORef (Map PackageLocationImmutable GenericPackageDescription))
+  , pcParsedCabalFilesRawImmutable :: !(IORef (Map RawPackageLocationImmutable GenericPackageDescription))
```
[the remark this answers is not in the document]
Comment (qrilka) [reply subs/pantry/src/Pantry/Types.hs]: It's needed for `loadCabalFileRawImmutable` which is currently used in Stack, if (or when) `Stack.Snapshot` will be remove we could get rid of "Raw" here. But if you have some suggestions how to handle it already now - please show

File: subs/pantry/src/Pantry/Types.hs [hunk 2]
```diff
@@ -1116,20 +1233,49 @@ instance Store OptionalSubdirs
 -- has the expected information.
```
[the remark this answers is not in the document]
Comment (qrilka) [reply subs/pantry/src/Pantry/Types.hs]: I'm not the author here but as far as I get it "config" here means either `stack.yaml` or `snapshot.yaml`. As for the package name field in it I don't know a good example when metadata could exist without a package name in it, maybe @snoyberg could explain it?

[the remark this answers is not in the document]
Comment (snoyberg) [reply subs/pantry/src/Pantry/Types.hs]: If you include a package as an archive or Git repo, you are not required to specify the name or version in the config, those are derived after parsing the cabal file.

File: subs/pantry/src/Pantry/Types.hs [hunk 3]
```diff
@@ -257,19 +285,54 @@ instance Display PackageLocation where
   display (PLImmutable loc) = display loc
   display (PLMutable fp) = fromString $ toFilePath $ resolvedAbsolute fp
 
+-- | Convert `PackageLocation` to its "raw" equivalent
+--
+-- @since 0.1.0.0
+toRawPL :: PackageLocation -> RawPackageLocation
+toRawPL (PLImmutable im) = RPLImmutable (toRawPLI im)
+toRawPL (PLMutable m) = RPLMutable m
```
Comment (NorfairKing) [inline subs/pantry/src/Pantry/Types.hs]: It's worth notice that the corresponding roundtrip loses info.
So if you start from a `RawPackageLocation`, then make it into a `PackageLocation` and then back, you'll lose information and that could make the error messages a bit annoying.
For error messages we should use `RawPackageLocation` if possible.

File: subs/pantry/src/Pantry/Types.hs [hunk 4]
```diff
@@ -257,19 +285,54 @@ instance Display PackageLocation where
...
+    (if T.null $ repoSubdir repo
+       then mempty
+       else " in subdir " <> display (repoSubdir repo))
+
 -- | Location for remote packages or archives assumed to be immutable.
 --
 -- @since 0.1.0.0
 data PackageLocationImmutable
-  = PLIHackage !PackageIdentifierRevision !(Maybe TreeKey)
+  = PLIHackage !PackageIdentifier !BlobKey !TreeKey
```
Comment (psibi) [inline subs/pantry/src/Pantry/Types.hs]: @qrilka Any reason why this is changed to `PackageIdentifier` from `PackageIdentifierRevision`. I see that in the Raw version it is `PackageIdentifierRevision`.

File: subs/pantry/src/Pantry/Types.hs (hunk 3 above)
Comment (qrilka) [reply subs/pantry/src/Pantry/Types.hs]: Nah, that opens up an interesting situation: if for a package in `extra-deps` you don't set a concrete revision Stack is supposed to resolve it and work with that resolved exact location (maybe from a lock file which is to appear soon). If we won't keep a corresponding raw location then error message could be misleading: e.g. after Stack lock file was created package author could have created a new revision fixing dependency problem but Stack will complain about that previous resolved revision if there will be no measures against such a situation.

File: src/Stack/Build/ConstructPlan.hs [hunk 5]
```diff
@@ -422,11 +423,16 @@ addDep treatAsDep' name = do
...
                             -- names. This code does not feel right.
-                            tellExecutablesUpstream
-                              name
-                              (PLIHackage (PackageIdentifierRevision name (installedVersion installed) CFILatest) Nothing)
-                              loc
-                              Map.empty
+                            let version = installedVersion installed
+                            mrev <- liftRIO $ getLatestHackageRevision name version
+                            case mrev of
+                              Nothing -> return () -- FIXME:qrilka report an error?
```
[the remark this answers is not in the document]
Comment (qrilka) [reply src/Stack/Build/ConstructPlan.hs]: Completely missed this, @psibi will you take care of it?

File: subs/pantry/src/Pantry/Types.hs (hunk 3 above)
Comment (NorfairKing) [reply subs/pantry/src/Pantry/Types.hs]: Aha! Thanks

File: src/Stack/Build/ConstructPlan.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (psibi) [reply src/Stack/Build/ConstructPlan.hs]: Sure.

File: subs/pantry/src/Pantry/Types.hs (hunk 4 above)
Comment (qrilka) [reply subs/pantry/src/Pantry/Types.hs]: for exact locations we don't need `CabalFileInfo`, , `TreeKey` is enough and e.g. `CFILatest` is not an exact revision :)

Comment (psibi) [reply subs/pantry/src/Pantry/Types.hs]: Ah, I see. Thanks.

**Comment by qrilka** [comment]
I think it's not quite correct to approve code which is mostly of my authorship but still overall I don't see any obvious issues with this PR
@psibi will you resolve the conflict?

**Comment by psibi** [comment]
@qrilka Sure, I will fix that by tomorrow.

**Comment by psibi** [comment]
@qrilka Resolved the conflicts and the CI is green too.

**Comment by mihaimaruseac** [comment]
Thanks to both @psibi and @qrilka for this

**Pull request merged**